addons/auth_totp_mail_enforce/models/res_users.py

```diff
@@ -0,0 +1,133 @@
+        ICP = self.env['ir.config_parameter'].sudo()
+        otp_required = False
+        if ICP.get_param('auth_totp.policy') == 'all_required':
+            otp_required = True
+        elif ICP.get_param('auth_totp.policy') == 'employee_required' and self.env.user.has_groups('base.group_user'):
+            otp_required = True
+        if otp_required:
+            return '/web/login/totp'
+
+    def _totp_check(self, code):
+        self._totp_rate_limit('code_check')
+        user = self.sudo()
+        if user.totp_enabled:
+            return super()._totp_check(code)
+
+        key = self._get_totp_mail_key()
+        match = TOTP(key).match(code, window=3600, timestep=3600)
+        if match is None:
+            _logger.info("2FA check (mail): FAIL for %s %r", self, self.login)
+            raise AccessDenied(_("Verification failed, please double-check the 6-digit code"))
+        _logger.info("2FA check(mail): SUCCESS for %s %r", self, self.login)
+        self._totp_rate_limit_purge('code_check')
+        return True
+
+    def _get_totp_mail_key(self):
+        self.ensure_one()
+        secret = self.env['ir.config_parameter'].sudo().get_param('database.secret')
+        assert secret, "Mail TOTP protection requires a configured database secret"
+        return f'{self.id}/{self.login}/{secret}'.encode()
+
+    def _get_totp_mail_code(self):
+        self.ensure_one()
+
+        key = self._get_totp_mail_key()
+
+        now = datetime.now()
+        counter = int(datetime.timestamp(now) / 3600)
+
+        code = hotp(key, counter)
+        expiration = now + timedelta(seconds=3600)
+
+        return code, expiration
+
+    def _send_totp_mail_code(self, user_demand=False):
+        self.ensure_one()
+        self._totp_rate_limit('send_email')
+
+        template = self.env.ref('auth_totp_mail_enforce.mail_template_totp_mail_code').sudo()
+        odoobot = self.env.ref('base.partner_root').sudo()
+        email_values = {
+            'email_from': odoobot.email_formatted,
+            'author_id': odoobot.id,
+        }
+        context = {}
+        if request:
+            geoip = request.session.geoip
+            environ = request.httprequest.environ
+            # e.g.
+            # from 'Mozilla/5.0 (Linux; Android 11; Pixel 5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.91'
+            # get 'Mozilla/5.0' as browser and 'Linux; Android 11; Pixel 5' as device
+            [(browser, device)] = re.findall('^([^ ]*)(?: \\(([^\\(]*)\\))?', environ['HTTP_USER_AGENT'])
+            context.update({
+                'location': f"{geoip['city']}, {geoip['country_name']}" if geoip else None,
+                'device': device,
+                'browser': browser,
+                'ip': environ['REMOTE_ADDR'],
+            })
+        return template.with_context(**context).send_mail(
+            self.id, force_send=True, email_values=email_values, notif_layout='mail.mail_notification_light'
+        )
```

Comment:
as a precaution, for emails containing sensitive data we do the following:
- `force_send=True` + `raise_exception=True` in order to immediately detect a wrong email or anything preventing delivery, and avoid having the email stay forever in the queue in "exception" state
- send inside a `savepoint()`, so that the failed email is cleanly rolled back from the database in case of failure
- force the `auto_delete`, `partner_ids`, `email_cc` etc. in `email_values` to avoid funny configurations where someone would permanently alter the template and compromise its safety

Cfr the reset password email as an example:
https://github.com/odoo/odoo/blob/15.0/addons/auth_signup/models/res_users.py#L187-L203

Reply:
Done